### DISCUSSION TITLE
[Documentation] "Channel-level exceptions" and channel recovery state

### DISCUSSION THREAD

**Discussion by HeathJared** [rabbitmq/rabbitmq-dotnet-client#1674] [Q&A] [answered]
### Is your feature request related to a problem? Please describe.

The RabbitMQ documentation frequently refers to "Channel-level exceptions" when discussing connection and channel recovery and describes several situations where these exceptions may need to be handled differently.

I am unable to find a list of exceptions that are considered "Channel-level exceptions" either through documentation or reviewing the source code for base classes, etc.

### Describe the solution you'd like

I would like any or all of the following:

- Documentation explicitly listing exceptions that are considered "Channel-level exceptions"
- A way to determine if a channel is closed, but still recoverable, versus a channel that is closed and un-recoverable.

### Describe alternatives you've considered

Add an `IsRecoverable` property to the `IModel` and handle updating it internally.

### Additional context

I have a use case where I am pooling channels for re-use by an ingestion API. It is my understanding that when a channel is closed due to a "Channel-level exception" that it is not recoverable and needs to be disposed instead of returning it to the pool. Meanwhile, channels that are closed, but recoverable, could still be returned to the pool to prevent churn (especially during a temporary loss of connection).

I currently have no way to reliably distinguish between these two situations.

**Comment by lukebakken** [comment]
> A way to determine if a channel is closed, but still recoverable, versus a channel that is closed and un-recoverable

If a channel is closed by RabbitMQ, it can _never_ be recovered.

**Reply by HeathJared** [reply]
Using the following settings on the connection:

```
AutomaticRecoveryEnabled = true,
TopologyRecoveryEnabled = true,
```

I do see that `channel.IsClosed` can be `true` while the connection is unavailable and after the connection and channel is recovered the `channel.IsClosed` becomes `false`.

However, as described in the original post, there are situations (Channel-level exceptions) that will prevent the channel from recovering.

How do I distinguish between the two situations?

**Reply by lukebakken** [reply]
Here is where connection recovery starts - https://github.com/rabbitmq/rabbitmq-dotnet-client/blob/6.x/projects/RabbitMQ.Client/client/impl/AutorecoveringConnection.cs#L709-L715

I don't have time to dig into this questions right now, but that's a good place for you to start. Nobody has asked for this kind of detail with regard to recovery so it's not documented, and I don't know it off the top of my head right now. I am busy finalizing the 7.0.0 release of this library, and supporting customers who pay for RabbitMQ. If I have time to return to this, I will, or maybe someone else in the community will chime in.

**Reply by HeathJared** [reply]
Thanks for your input and work on the client. I will take another look.

**Reply by lukebakken** [reply]
For automatic recovery, channels are always recovered, but they are not the _same_ channel within RabbitMQ. This library re-uses the `IModel` instance for the new channel in RabbitMQ, and updates its internal state for the new RabbitMQ channel. This is transparent to the user. So, you should be able to return your `IModel` instances to the pool all the time.

So, if you find a situation that doesn't work as expected, please let me know.

**Reply by HeathJared** [reply]
If that is the case, that is good news. I will let it run under stress for a few weeks and let you know if I encounter a channel that did not recover as expected.

Most of my understanding was based strictly on the statements from the RabbitMQ documentation. If this library handles the problem automatically that is excellent.

Thanks and good luck on 7.0.

**Reply by lukebakken** [reply, marked as answer]
If you have time, I suggest migrating your code to use the latest version 7 RC - https://www.nuget.org/packages/RabbitMQ.Client/7.0.0-rc.11

I intended to ship that version today but there are some small API changes coming. Version 7 works much better than version 6 in pretty much every way, especially in its use of the thread pool.

**Reply by HeathJared** [reply]
I hope to soon. We won't be able to this release cycle, but its on the list to revisit soon (Q4 hopefully) along with some other changes and optimizations of our usage of RabbitMQ. Looking forward to it.

**Reply by HeathJared** [reply]
I wanted to stop by and say good job on the release. The improvements are very much noticed.

I've been stress testing v7 so we can finally migrate from v6 for production and we have been able to remove our channel pools, remove our async confirmation logic, and increase our message throughput anywhere from 30% to 200% depending on the situation. Everything appears to recover as expected.

Benchmarks also greatly improved. For example:

| Method           | Mean     | Error   | StdDev  |
|----------------- |---------:|--------:|--------:|
| PublishAsync_v6_Net8 | 217.6 us | 4.02 us | 5.76 us |

| Method           | Mean     | Error    | StdDev   |
|----------------- |---------:|---------:|---------:|
| PublishAsync_v7_Net9 | 57.08 ns | 0.232 ns | 0.206 ns |


I'm currently watching this [issue](https://github.com/rabbitmq/rabbitmq-dotnet-client/issues/1818) while we work towards the next production release including v7.

Just wanted to say thanks again.

**Reply by lukebakken** [reply]
@HeathJared have you seen anything like what is reported in #1818? I would be very interested to know!

**Reply by HeathJared** [reply]
No we have not encountered that (yet). Our tests and use cases maybe different. 

One of our primary use cases being tested is publishing to channels from multiple threads (high traffic ingestion API with lots of competing publishers) with publish confirmation, so I believe it to be a very similar situation.

If I do encounter this and the ticket is not yet resolved, I plan to add any findings that we dig up.

**Reply by lukebakken** [reply]
Thanks. I think the person who reported #1818 is experiencing a lot of `basic.nack` / `basic.return` from RabbitMQ, resulting in `PublishException` being thrown. Plus, they also said they have timeouts waiting for confirmations from RabbitMQ, which indicates the server is overloaded.

**Reply by HeathJared** [reply]
I did see that. 

We are using the default timeouts.
We are currently using RabbitMQ 4.0.5.
We do have a `catch` for `PublishException` as well to watch for failed confirmations.
We are also using `mandatory` `true`.

I do know that our publishing is spiky once we max out the load. It will peak, drop to zero, peak, drop to zero. 
I suspect this is due to the rate limiting, but we have not investigated yet.
